src/v/io/persistence.cc

```diff
@@ -0,0 +1,327 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#include "io/persistence.h"
+
+#include "units.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/core/seastar.hh>
+#include <seastar/util/later.hh>
+
+#include <span>
+
+namespace {
+/*
+ * build an alignment error to return from read/write
+ */
+seastar::future<size_t> make_alignment_error(
+  std::string_view op, std::string_view arg, auto val, uint64_t alignment) {
+    auto msg = fmt::format(
+      R"(Op "{}" arg "{}" with value {} expects alignment {})",
+      op,
+      arg,
+      val,
+      alignment);
+    return seastar::make_exception_future<size_t>(
+      std::system_error(std::error_code(EINVAL, std::system_category()), msg));
+}
+
+/*
+ * helper for checking alignment of read/write
+ */
+std::optional<seastar::future<size_t>> check_alignment(
+  std::string_view name,
+  uint64_t pos,
+  const char* buf,
+  size_t len, // NOLINT(bugprone-easily-swappable-parameters)
```

Comment:
nit: should this be a `std::span` to remove the clang-tidy warning? I'm assuming the `len` is for `buf`?

Reply:
yeh, good idea. it works, but looks weird. i think i'm just not used to std::span yet.